File: rust/core/src/timeline/placement/insert_index.rs
//! Computes the correct insertion index for new tracks based on
//! track type, preferred position, and directional hints ("above"/"below").

use crate::timeline::models::SceneTracks;

/// Returns the default insertion index for a new track of the given type.
///
/// - Audio tracks are placed after the main track and all overlay tracks.
/// - Effect tracks are placed at index 0 (top).
/// - All other track types are placed just above the main track.
pub fn get_default_insert_index_for_track(tracks: &SceneTracks, track_type: &str) -> usize {
    if track_type == "audio" {
        return tracks.overlay.len() + 1 + tracks.audio.len();
    }

    if track_type == "effect" {
        return 0;
    }

    tracks.overlay.len()
}

/// Returns the highest (topmost) possible insertion index for a new track
/// of the given type. For audio tracks this is immediately after the main
/// track; all other types return index 0.
pub fn get_highest_insert_index_for_track(tracks: &SceneTracks, track_type: &str) -> usize {
    if track_type == "audio" {
        return tracks.overlay.len() + 1;
    }

    0
}

/// The computed placement for a new track: its insert index and an optional
/// positional hint ("above" or "below").
pub struct PreferredNewTrackPlacement {
    pub insert_index: usize,
    pub insert_position: Option<String>,
}

/// Resolves the optimal position for a new track, taking into account the
/// current track layout, the preferred index, and the placement direction
/// ("above" or "below").
pub fn resolve_preferred_new_track_placement(
    tracks: &SceneTracks,
    track_type: &str,
    preferred_index: usize,
    direction: &str, // "above" | "below"
) -> PreferredNewTrackPlacement {
    let track_count = tracks.overlay.len() + 1 + tracks.audio.len();

    if track_count == 0 {
        return PreferredNewTrackPlacement {
            insert_index: 0,
            insert_position: if track_type == "audio" {
                Some("below".to_string())
            } else {
                None
            },
        };
    }

    let safe_preferred_index = preferred_index.min(track_count.saturating_sub(1));
    let main_track_index = tracks.overlay.len();

    if track_type == "audio" {
        if safe_preferred_index <= main_track_index {
            return PreferredNewTrackPlacement {
                insert_index: main_track_index + 1,
                insert_position: Some("below".to_string()),
            };
        }

        return PreferredNewTrackPlacement {
            insert_index: if direction == "above" {
                safe_preferred_index
            } else {
                safe_preferred_index + 1
            },
            insert_position: Some(direction.to_string()),
        };
    }

    let insert_index = if direction == "above" {
        safe_preferred_index
    } else {
        safe_preferred_index + 1
    };

    if insert_index > main_track_index {
        return PreferredNewTrackPlacement {
            insert_index: main_track_index,
            insert_position: Some("above".to_string()),
        };
    }

    PreferredNewTrackPlacement {
        insert_index,
        insert_position: Some(direction.to_string()),
    }
}

File: apps/web/src/timeline/placement/insert-index.ts
/**
 * Insert-index helpers for placing new tracks within the overlay/audio
 * ordering, respecting the main track position.
 *
 * @module timeline/placement/insert-index
 */

import type { SceneTracks, TrackType } from "@/timeline";

/**
 * Returns the default insert index for a new track of the given type.
 */
export function getDefaultInsertIndexForTrack({
	tracks,
	trackType,
}: {
	tracks: SceneTracks;
	trackType: TrackType;
}): number {
	if (trackType === "audio") {
		return tracks.overlay.length + 1 + tracks.audio.length;
	}

	if (trackType === "effect") {
		return 0;
	}

	return tracks.overlay.length;
}

/**
 * Returns the highest (farthest right) insert index for a new track.
 */
export function getHighestInsertIndexForTrack({
	tracks,
	trackType,
}: {
	tracks: SceneTracks;
	trackType: TrackType;
}): number {
	if (trackType === "audio") {
		return tracks.overlay.length + 1;
	}

	return 0;
}

/**
 * Resolves the insert index and position (above/below) for a new track,
 * keeping audio tracks always below the main track.
 */
export function resolvePreferredNewTrackPlacement({
	tracks,
	trackType,
	preferredIndex,
	direction,
}: {
	tracks: SceneTracks;
	trackType: TrackType;
	preferredIndex: number;
	direction: "above" | "below";
}): { insertIndex: number; insertPosition: "above" | "below" | null } {
	const trackCount = tracks.overlay.length + 1 + tracks.audio.length;
	if (trackCount === 0) {
		return {
			insertIndex: 0,
			insertPosition: trackType === "audio" ? "below" : null,
		};
	}

	const safePreferredIndex = Math.min(
		Math.max(preferredIndex, 0),
		trackCount - 1,
	);
	const mainTrackIndex = tracks.overlay.length;

	if (trackType === "audio") {
		if (safePreferredIndex <= mainTrackIndex) {
			return {
				insertIndex: mainTrackIndex + 1,
				insertPosition: "below",
			};
		}

		return {
			insertIndex:
				direction === "above" ? safePreferredIndex : safePreferredIndex + 1,
			insertPosition: direction,
		};
	}

	const insertIndex =
		direction === "above" ? safePreferredIndex : safePreferredIndex + 1;
	if (mainTrackIndex >= 0 && insertIndex > mainTrackIndex) {
		return {
			insertIndex: mainTrackIndex,
			insertPosition: "above",
		};
	}

	return {
		insertIndex,
		insertPosition: direction,
	};
}
